### PR TITLE
feat(atomic): atomic-sort-dropdown completion

### PR DESCRIPTION
**Pull request by ThibodeauJF**
Tests coming in a separate PR

Screen Shot 2021-02-02 at 6 18 47 PM

Screen Shot 2021-02-02 at 6 18 53 PM

https://coveord.atlassian.net/browse/KIT-394

### CODE REVIEW

File: packages/atomic/package.json [hunk 1]
```diff
@@ -48,6 +48,7 @@
     "@types/mustache": "^4.0.1",
     "@types/postcss-import": "^12.0.0",
     "@types/puppeteer": "5.4.0",
+    "autoprefixer": "^10.2.4",
```
Comment (ThibodeauJF) [inline packages/atomic/package.json:51]: https://github.com/tailwindlabs/tailwindcss/issues/23#issuecomment-341136753

**Review by btaillon — APPROVED** [review]

**Review by samisayegh — APPROVED** [review]
How would the current design support composite criteria? Or is it not commonly used?

File: packages/atomic/src/components/atomic-sort-criterion/atomic-sort-criterion.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+@Component({
+  tag: 'atomic-sort-criterion',
+  shadow: true,
+})
+export class AtomicSortCriterion {
+  @Element() host!: HTMLElement;
+  /**
+   * The available criteria that can be used to sort query results.
+   *
+   * - `relevancy`: Uses standard index ranking factors (adjacency, TDIDF) and custom ranking expressions (QREs and QRFs) to compute a ranking score for each query result item, and sorts the query results by descending score value.
```
[the remark this answers is not in the document]
Comment (olamothe) [reply packages/atomic/src/components/atomic-sort-criterion/atomic-sort-criterion.tsx:22]: It's `TFIDF` :) 

https://monkeylearn.com/blog/what-is-tf-idf/#:~:text=TF%2DIDF%20is%20a%20statistical,across%20a%20set%20of%20documents.

File: packages/headless/src/features/sort-criteria/criteria.ts [hunk 3]
```diff
@@ -135,3 +141,53 @@ export const buildQueryRankingExpressionSortCriterion = (): SortByQRE => ({
...
+    field: new StringValue(),
+  },
+});
+
+/**
+ * Utility function that validates and builds a valid criterion. Throws an error when the criterion is invalid.
+ * @param criterion The criterion definition.
+ * @returns A criterion.
+ */
+export function validateSortCriterion(criterion: {
```
[the remark this answers is not in the document]
Comment (ThibodeauJF) [reply packages/headless/src/features/sort-criteria/criteria.ts:158]: The validator validates a single, not hardcoded, Criterion at a time. I'm not using the expression Either but keeping the objects. I'll see if I'll switch to expression if I end up supporting composites

File: packages/atomic/src/components/atomic-sort-criterion/atomic-sort-criterion.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  @Prop({reflect: true}) field?: string;
+
+  /**
+   * The non-localized caption to display for this criteria.
+   */
+  @Prop({mutable: true, reflect: true}) caption!: string;
+
+  /**
+   * Validated sort criterion. `null` when invalid.
+   * For development purposes, not meant to be specified by users.
```
[the remark this answers is not in the document]
Comment (olamothe) [reply packages/atomic/src/components/atomic-sort-criterion/atomic-sort-criterion.tsx:46]: Indeed, I'm not sure I understand why it needs to be a `@Prop`. 

Can't it be a simple class member, not externally available ?

**Review by olamothe — APPROVED** [review]
As discussed in Slack, we do indeed need to support composite sort.

I'll approve this PR if you want to either merge and this and work on the support in a new one, otherwise please ping me if you end up doing the change as part of this one.

**Review by nrawji — APPROVED** [review]

File: packages/atomic/src/components/atomic-sort-criterion/atomic-sort-criterion.tsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+  /**
+   * The sortable field on which to sort when `by` is `field`.
+   */
+  @Prop({reflect: true}) field?: string;
+
+  /**
+   * The non-localized caption to display for this criteria.
+   */
+  @Prop({mutable: true, reflect: true}) caption!: string;
```
[the remark this answers is not in the document]
Comment (ThibodeauJF) [reply packages/atomic/src/components/atomic-sort-criterion/atomic-sort-criterion.tsx:42]: theres a way, you can see a commented snippet in index.html, line 26
```
searchInterface.i18n.addResource('en', 'translation', 'search', 'Make me feel lucky!');
```

File: packages/atomic/cypress/integration/sort-dropdown.cypress.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      </atomic-sort-dropdown>
+    `);
+    cy.wait(500);
+  }
+
+  function selectOption(value: string) {
+    return cy.get(sortDropdown).shadow().find('select').select(value);
+  }
+
+  function shouldRenderErrorComponent() {
```
Comment (lvu285) [inline packages/atomic/cypress/integration/sort-dropdown.cypress.ts:22]: can we reuse `shouldRenderErrorComponent` function that I created

File: packages/atomic/cypress/integration/sort-dropdown.cypress.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
+import {getApiRequestBody, getAnalytics} from '../utils/network';
+import {setUpPage} from '../utils/setupComponent';
+
+const sortDropdown = 'atomic-sort-dropdown';
+const criteria = ['relevancy', 'date descending'];
```
Comment (lvu285) [inline packages/atomic/cypress/integration/sort-dropdown.cypress.ts:5]: I see that the requirement for criteria is `criteria: criteria expression, can be multiple criteria, separated by commas`,
so can we have a test that criteria contains more than 1 field?

**Pull request merged**